stdlib/json.gr

```diff
@@ -0,0 +1,2112 @@
+   *
+   * ```json
+   * []
+   * ```
+   *
+   * ```json
+   * [↵
+   * ··1↵
+   * ]
+   * ```
+   *
+   * ```json
+   * [↵
+   * ··1,↵
+   * ··2,↵
+   * ··3↵
+   * ]
+   * ```
+   */
+  OneArrayEntryPerLine,
+}
+
+/**
+ * Controls how objects are printed in custom formatting.
+ *
+ * The following examples have whitespaces and line breaks replaced with visible
+ * characters for illustrative purposes.
+ */
+provide enum ObjectFormat {
+  /**
+   * Objects are emitted in a compact manner.
+   *
+   * ```json
+   * {}
+   * ```
+   *
+   * ```json
+   * {"a":1}
+   * ```
+   *
+   * ```json
+   * {"a":1,"b":2,"c":3}
+   * ```
+   */
+  CompactObjectEntries,
+  /**
+   * Objects are emitted with spaces between entries.
+   *
+   * ```json
+   * { }
+   * ```
+   *
+   * ```json
+   * {"a": 1}
+   * ```
+   *
+   * ```json
+   * {"a": 1, "b": 2, "c": 3}
+   * ```
+   */
+  SpacedObjectEntries,
+  /**
+   * Objects are emitted with each entry on a new line.
+   *
+   * ```
+   * {}
+   * ```
+   *
+   * ```
+   * {↵
+   * ··"a":·1↵
+   * }
+   * ```
+   *
+   * ```
+   * {↵
+   * ··"a":·1,↵
+   * ··"b":·2,↵
+   * ··"c":·3↵
+   * }
+   * ```
+   */
+  OneObjectEntryPerLine,
+}
+
+/**
+ * Controls line ending type in custom formatting.
+ */
+provide enum LineEnding {
+  /**
+   * No line endings will be emitted.
+   */
+  NoLineEnding,
+  /**
+   * A `\n` will be emitted at the end of each line.
+   */
+  LineFeed,
+  /**
+   * A `\r\n` will be emitted at the end of each line
+   */
+  CarriageReturnLineFeed,
+  /**
+   * A `\r` will be emitted at the end of each line
+   */
+  CarriageReturn,
+}
+
+/*
+ * Allows fine-grained control of formatting in JSON printing.
+ */
+record FormattingSettings {
+  indentation: IndentationFormat,
+  arrayFormat: ArrayFormat,
+  objectFormat: ObjectFormat,
+  lineEnding: LineEnding,
+  finishWithNewLine: Bool,
+  escapeAllControlPoints: Bool,
+  escapeHTMLUnsafeSequences: Bool,
+  escapeNonASCII: Bool,
+}
+
+/**
+ * Allows control of formatting in JSON printing.
+ */
+provide enum FormattingChoices {
+  /**
+   * Recommended human readable formatting.
+   *
+   * Escapes all control points for the sake of clarity, but prints unicode
+   * codepoints directly so the result needs to be treated as proper unicode and
+   * is not safe to be transported in ASCII encoding.
+   *
+   * Roughly Equivalent to:
+   * ```grain
+   * Custom{
+   *  indentation: IndentWithSpaces(2),
+   *  arrayFormat: OneArrayEntryPerLine,
+   *  objectFormat: OneObjectEntryPerLine,
+   *  lineEnding: LineFeed,
+   *  finishWithNewLine: true,
+   *  escapeAllControlPoints: true,
+   *  escapeHTMLUnsafeSequences: false,
+   *  escapeNonASCII: false,
+   * }
+   * ```
+   *
+   * The following example have whitespaces, line breaks and control points
+   * replaced with visible characters.
+   * ```json
+   * {↵
+   * ··"currency":·"€",↵
+   * ··"price":·99.9,↵
+   * ··"currencyDescription":·"EURO\u007f",↵
+   * }
+   * ```
```

Comment:
I don't think these json examples are necessary here, but maybe @ospencer feels differently. If we keep them, we need to streamline like I mentioned above, as I dislike the syntax highlighting.

Reply:
I removed the illustrative whitespace but I do still think that the examples add value.

Reply:
It's fine to keep them, but we need to remove the dots, as I've commented in my follow up PR. We should be using tabs and spaces as the actual output would